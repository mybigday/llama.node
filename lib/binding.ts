import * as path from "path";

export type ChatMessage = {
  role: string;
  content: string;
};

export type LlamaModelOptions = {
  model: string;
  chat_template?: string;
  embedding?: boolean;
  embd_normalize?: number;
  pooling_type?: "none" | "mean" | "cls" | "last" | "rank";
  n_ctx?: number;
  n_batch?: number;
  n_ubatch?: number;
  n_threads?: number;
  n_gpu_layers?: number;
  flash_attn?: boolean;
  cache_type_k?:
    | "f16"
    | "f32"
    | "q8_0"
    | "q4_0"
    | "q4_1"
    | "iq4_nl"
    | "q5_0"
    | "q5_1";
  cache_type_v?:
    | "f16"
    | "f32"
    | "q8_0"
    | "q4_0"
    | "q4_1"
    | "iq4_nl"
    | "q5_0"
    | "q5_1";
  use_mlock?: boolean;
  use_mmap?: boolean;
  vocab_only?: boolean;
  lora?: string;
  lora_scaled?: number;
  lora_list?: { path: string; scaled: number }[];
};

export type CompletionResponseFormat = {
  type: "text" | "json_object" | "json_schema";
  json_schema?: {
    strict?: boolean;
    schema: object;
  };
  schema?: object; // for json_object type
};

export type LlamaCompletionOptions = {
  messages?: ChatMessage[];
  jinja?: boolean;
  chat_template?: string;
  response_format?: CompletionResponseFormat;
  tools?: object;
  parallel_tool_calls?: boolean;
  tool_choice?: string;
  prompt?: string;
  temperature?: number;
  top_k?: number;
  top_p?: number;
  min_p?: number;
  mirostat?: number;
  mirostat_tau?: number;
  mirostat_eta?: number;
  penalty_last_n?: number;
  penalty_repeat?: number;
  penalty_freq?: number;
  penalty_present?: number;
  typ_p?: number;
  xtc_threshold?: number;
  xtc_probability?: number;
  dry_multiplier?: number;
  dry_base?: number;
  dry_allowed_length?: number;
  dry_penalty_last_n?: number;
  n_predict?: number;
  max_length?: number;
  max_tokens?: number;
  seed?: number;
  stop?: string[];
  grammar?: string;
  grammar_lazy?: boolean;
  grammar_triggers?: { word: string; at_start: boolean }[];
  preserved_tokens?: string[];
};

export type LlamaCompletionResult = {
  text: string;
  tokens_predicted: number;
  tokens_evaluated: number;
  truncated: boolean;
  timings: {
    prompt_n: number;
    prompt_ms: number;
    prompt_per_token_ms: number;
    prompt_per_second: number;
    predicted_n: number;
    predicted_ms: number;
    predicted_per_token_ms: number;
    predicted_per_second: number;
  };
};

export type LlamaCompletionToken = {
  token: string;
};

export type TokenizeResult = {
  tokens: Int32Array;
};

export type EmbeddingResult = {
  embedding: Float32Array;
};

export interface LlamaContext {
  new (options: LlamaModelOptions): LlamaContext;
  getSystemInfo(): string;
  getModelInfo(): object;
  getFormattedChat(
    messages: ChatMessage[],
    chat_template?: string,
    params?: {
      jinja?: boolean;
      response_format?: CompletionResponseFormat;
      tools?: object;
      parallel_tool_calls?: object;
      tool_choice?: string;
    }
  ): object | string;
  completion(
    options: LlamaCompletionOptions,
    callback?: (token: LlamaCompletionToken) => void
  ): Promise<LlamaCompletionResult>;
  stopCompletion(): void;
  tokenize(text: string): Promise<TokenizeResult>;
  detokenize(tokens: number[]): Promise<string>;
  embedding(text: string): Promise<EmbeddingResult>;
  saveSession(path: string): Promise<void>;
  loadSession(path: string): Promise<void>;
  release(): Promise<void>;
  applyLoraAdapters(adapters: { path: string; scaled: number }[]): void;
  removeLoraAdapters(adapters: { path: string }[]): void;
  getLoadedLoraAdapters(): { path: string; scaled: number }[];
  // static
  loadModelInfo(path: string, skip: string[]): Promise<Object>;
}

export interface Module {
  LlamaContext: LlamaContext;
}

export type LibVariant = "default" | "vulkan" | "cuda";

const setupEnv = (variant?: string) => {
  const postfix = variant ? `-${variant}` : "";
  const binPath = path.resolve(
    __dirname,
    `../bin/${process.platform}${postfix}/${process.arch}/`
  );
  const systemPathEnv = process.env.PATH ?? process.env.Path ?? "";
  if (!systemPathEnv.includes(binPath)) {
    if (process.platform === "win32") {
      process.env.Path = `${binPath};${systemPathEnv}`;
    } else {
      process.env.PATH = `${binPath}:${systemPathEnv}`;
    }
  }
};

export const loadModule = async (variant?: LibVariant): Promise<Module> => {
  try {
    if (variant && variant !== "default") {
      setupEnv(variant);
      return (await import(
        `../bin/${process.platform}-${variant}/${process.arch}/llama-node.node`
      )) as Module;
    }
  } catch {} // ignore errors and try the common path
  setupEnv();
  return (await import(
    `../bin/${process.platform}/${process.arch}/llama-node.node`
  )) as Module;
};
