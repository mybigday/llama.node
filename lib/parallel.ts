// Parallel decoding API implementation for llama.node
import type {
  LlamaContext,
  LlamaCompletionOptions,
  LlamaCompletionToken,
  RerankParams,
} from './binding'

export class LlamaParallelAPI {
  private context: LlamaContext
  private enabled: boolean = false
  private pendingRequests = new Map<
    number,
    {
      resolve: (value: any) => void
      reject: (reason?: any) => void
    }
  >()

  constructor(context: LlamaContext) {
    this.context = context
  }

  /**
   * Enable parallel decoding mode
   * @param config Configuration for parallel mode
   * @returns boolean indicating if successful
   */
  async enable(config?: {
    n_parallel?: number
    n_batch?: number
  }): Promise<boolean> {
    const defaultConfig = { n_parallel: 2, n_batch: 512 }
    const result = this.context.enableParallelMode({
      ...defaultConfig,
      ...config,
    })
    this.enabled = result
    return result
  }

  /**
   * Disable parallel decoding mode
   */
  disable(): void {
    this.context.disableParallelMode()
    this.enabled = false
  }

  /**
   * Configure parallel decoding mode (enables if not already enabled)
   * @param config Configuration for parallel mode
   * @returns boolean indicating if successful
   */
  async configure(config: {
    n_parallel?: number
    n_batch?: number
  }): Promise<boolean> {
    return this.enable(config)
  }

  /**
   * Queue a completion request for parallel processing
   * @param options Completion options
   * @param onToken Optional callback for each token
   * @returns Object with requestId, promise for result, and stop function
   */
  async completion(
    options: LlamaCompletionOptions,
    onToken?: (requestId: number, data: LlamaCompletionToken) => void,
  ): Promise<{
    requestId: number
    promise: Promise<any>
    stop: () => void
  }> {
    if (!this.enabled) {
      throw new Error('Parallel mode is not enabled. Call enable() first.')
    }

    const tokenCallback = onToken
      ? (error: any, result: any) => {
          if (error) {
            console.error('Token callback error:', error)
            // Handle completion error
            const pendingReq = this.pendingRequests.get(result?.requestId)
            if (pendingReq) {
              pendingReq.reject(error)
              this.pendingRequests.delete(result?.requestId)
            }
            return
          }
          // Check if this is a token callback or final result
          if (result) {
            if (result.token !== undefined) {
              // This is a token callback
              onToken(result.requestId, result)
            } else if (
              result.text !== undefined ||
              result.content !== undefined
            ) {
              // This is the final result
              const pendingReq = this.pendingRequests.get(result.requestId)
              if (pendingReq) {
                pendingReq.resolve(result)
                this.pendingRequests.delete(result.requestId)
              }
            }
          }
        }
      : undefined

    // Queue the completion immediately (this is synchronous!)
    const { requestId } = this.context.queueCompletion(
      options,
      tokenCallback ||
        ((error, result) => {
          if (error) {
            const pendingReq = this.pendingRequests.get(result?.requestId)
            if (pendingReq) {
              pendingReq.reject(error)
              this.pendingRequests.delete(result?.requestId)
            }
          } else if (
            result &&
            (result.text !== undefined || result.content !== undefined)
          ) {
            // Final result for non-streaming
            const pendingReq = this.pendingRequests.get(result.requestId)
            if (pendingReq) {
              pendingReq.resolve(result)
              this.pendingRequests.delete(result.requestId)
            }
          }
        }),
    )

    // Create promise for final result
    const promise = new Promise((resolveResult, rejectResult) => {
      this.pendingRequests.set(requestId, {
        resolve: resolveResult,
        reject: rejectResult,
      })
    })

    // Create stop function
    const stop = () => {
      this.context.cancelRequest(requestId)
      const pendingReq = this.pendingRequests.get(requestId)
      if (pendingReq) {
        pendingReq.reject(new Error('Request cancelled'))
        this.pendingRequests.delete(requestId)
      }
    }

    // Return immediately without wrapping in a Promise
    return {
      requestId,
      promise,
      stop,
    }
  }

  /**
   * Queue an embedding request for parallel processing
   * @param text Text to embed
   * @param params Optional embedding parameters
   * @returns Object with requestId and promise for result
   */
  async embedding(
    text: string,
    params?: { embd_normalize?: number },
  ): Promise<{
    requestId: number
    promise: Promise<{ embedding: number[] }>
  }> {
    if (!this.enabled) {
      throw new Error('Parallel mode is not enabled. Call enable() first.')
    }

    // Create promise for result
    let resolveResult: (value: any) => void
    let rejectResult: (reason?: any) => void

    const promise = new Promise<{ embedding: number[] }>((res, rej) => {
      resolveResult = res
      rejectResult = rej
    })

    // Queue the embedding immediately (this is synchronous!)
    const { requestId } = this.context.queueEmbedding(
      text,
      params,
      (error, result) => {
        if (error) {
          rejectResult(error)
        } else {
          resolveResult(result)
        }
      },
    )

    // Return immediately without wrapping in a Promise
    return {
      requestId,
      promise,
    }
  }

  /**
   * Queue a rerank request for parallel processing
   * @param query Query text
   * @param documents Documents to rank
   * @param params Optional rerank parameters
   * @returns Object with requestId and promise for results
   */
  async rerank(
    query: string,
    documents: string[],
    params?: RerankParams,
  ): Promise<{
    requestId: number
    promise: Promise<Array<{ score: number; index: number; document: string }>>
  }> {
    if (!this.enabled) {
      throw new Error('Parallel mode is not enabled. Call enable() first.')
    }

    // Create promise for result
    let resolveResult: (value: any) => void
    let rejectResult: (reason?: any) => void

    const promise = new Promise<
      Array<{ score: number; index: number; document: string }>
    >((res, rej) => {
      resolveResult = res
      rejectResult = rej
    })

    // Queue the rerank immediately (this is synchronous!)
    const { requestId } = this.context.queueRerank(
      query,
      documents,
      params,
      (error, result) => {
        if (error) {
          rejectResult(error)
        } else {
          // Add document text to results and sort by score
          const enrichedResults = result.results
            .map((r: any) => ({
              ...r,
              document: documents[r.index],
            }))
            .sort((a: any, b: any) => b.score - a.score)
          resolveResult(enrichedResults)
        }
      },
    )

    // Return immediately without wrapping in a Promise
    return {
      requestId,
      promise,
    }
  }

  /**
   * Check if parallel mode is enabled
   */
  isEnabled(): boolean {
    return this.enabled
  }
}
