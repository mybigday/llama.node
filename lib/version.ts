export const BUILD_NUMBER = '6096';
export const BUILD_COMMIT = 'fd1234cb';
