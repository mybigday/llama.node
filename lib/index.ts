import { loadModule, LlamaModelOptions } from './binding'
import type {
  Module,
  LlamaContext,
  LibVariant,
  ChatMessage,
  LlamaCompletionOptions,
  LlamaCompletionToken,
  LlamaCompletionResult,
  TokenizeResult,
  EmbeddingResult,
  RerankParams,
  RerankResult,
  CompletionResponseFormat,
  ModelInfo,
  JinjaFormattedChatResult,
  Tool,
  GGUFModelInfo,
} from './binding'
import { BUILD_NUMBER, BUILD_COMMIT } from './version'
import { LlamaParallelAPI } from './parallel'

export * from './binding'
export { LlamaParallelAPI }

export const MTMD_DEFAULT_MEDIA_MARKER = '<__media__>'

export interface LlamaModelOptionsExtended extends LlamaModelOptions {
  lib_variant?: LibVariant
}

const mods: { [key: string]: Module } = {}

const logListeners: Array<(level: string, text: string) => void> = []

const logCallback = (level: string, text: string) => {
  logListeners.forEach((listener) => listener(level, text))
}

let logEnabled = false

const refreshNativeLogSetup = () => {
  Object.entries(mods).forEach(([, mod]) => {
    mod.LlamaContext.toggleNativeLog(logEnabled, logCallback)
  })
}

export const toggleNativeLog = async (enable: boolean) => {
  logEnabled = enable
  refreshNativeLogSetup()
}

export function addNativeLogListener(
  listener: (level: string, text: string) => void,
): { remove: () => void } {
  logListeners.push(listener)
  return {
    remove: () => {
      logListeners.splice(logListeners.indexOf(listener), 1)
    },
  }
}

const getJsonSchema = (responseFormat?: CompletionResponseFormat) => {
  if (responseFormat?.type === 'json_schema') {
    return responseFormat.json_schema?.schema
  }
  if (responseFormat?.type === 'json_object') {
    return responseFormat.schema || {}
  }
  return null
}

export type FormattedChatResult = {
  type: 'jinja' | 'llama-chat'
  prompt: string
  has_media: boolean
  media_paths?: Array<string>
}

class LlamaContextWrapper {
  ctx: LlamaContext
  parallel: LlamaParallelAPI

  constructor(nativeCtx: LlamaContext) {
    this.ctx = nativeCtx
    this.parallel = new LlamaParallelAPI(nativeCtx)
  }

  getSystemInfo(): string {
    return this.ctx.getSystemInfo()
  }

  getModelInfo(): ModelInfo {
    return this.ctx.getModelInfo()
  }

  isJinjaSupported(): boolean {
    const { minja } = this.ctx.getModelInfo().chatTemplates
    return !!minja?.toolUse || !!minja?.default
  }

  isLlamaChatSupported(): boolean {
    return !!this.ctx.getModelInfo().chatTemplates.llamaChat
  }

  _formatMediaChat(messages: ChatMessage[] | undefined): {
    messages: ChatMessage[] | undefined
    has_media: boolean
    media_paths?: string[]
  } {
    if (!messages)
      return {
        messages,
        has_media: false,
      }
    const mediaPaths: string[] = []
    return {
      messages: messages.map((msg) => {
        if (Array.isArray(msg.content)) {
          const content = msg.content.map((part) => {
            // Handle multimodal content
            if (part.type === 'image_url') {
              let path = part.image_url?.url || ''
              mediaPaths.push(path)
              return {
                type: 'text',
                text: MTMD_DEFAULT_MEDIA_MARKER,
              }
            } else if (part.type === 'input_audio') {
              const { input_audio: audio } = part
              if (!audio) throw new Error('input_audio is required')

              const { format } = audio
              if (format != 'wav' && format != 'mp3') {
                throw new Error(`Unsupported audio format: ${format}`)
              }
              if (audio.url) {
                const path = audio.url.replace(/file:\/\//, '')
                mediaPaths.push(path)
              } else if (audio.data) {
                mediaPaths.push(audio.data)
              }
              return {
                type: 'text',
                text: MTMD_DEFAULT_MEDIA_MARKER,
              }
            }
            return part
          })

          return {
            ...msg,
            content,
          }
        }
        return msg
      }),
      has_media: mediaPaths.length > 0,
      media_paths: mediaPaths,
    }
  }

  getFormattedChat(
    messages: ChatMessage[],
    template?: string,
    params?: {
      jinja?: boolean
      response_format?: CompletionResponseFormat
      tools?: Tool[]
      parallel_tool_calls?: boolean
      tool_choice?: string
      enable_thinking?: boolean
      add_generation_prompt?: boolean
      now?: string | number
      chat_template_kwargs?: Record<string, string>
    },
  ): FormattedChatResult {
    const {
      messages: chat,
      has_media,
      media_paths,
    } = this._formatMediaChat(messages)

    const useJinja = this.isJinjaSupported() && params?.jinja
    let tmpl
    if (template) tmpl = template // Force replace if provided

    const result = this.ctx.getFormattedChat(chat!, tmpl, {
      jinja: useJinja,
      response_format: params?.response_format,
      tools: params?.tools,
      parallel_tool_calls: params?.parallel_tool_calls,
      tool_choice: params?.tool_choice,
      enable_thinking: params?.enable_thinking ?? true,
      add_generation_prompt: params?.add_generation_prompt,
      now: params?.now,
      chat_template_kwargs: params?.chat_template_kwargs
        ? Object.entries(params.chat_template_kwargs).reduce(
            (acc, [key, value]) => {
              acc[key] = JSON.stringify(value) // Each value is a stringified JSON object
              return acc
            },
            {} as Record<string, any>,
          )
        : undefined,
    })

    if (!useJinja) {
      return {
        type: 'llama-chat',
        prompt: result as string,
        has_media,
        media_paths,
      }
    }
    const jinjaResult = result as JinjaFormattedChatResult
    return {
      type: 'jinja',
      has_media,
      media_paths,
      ...jinjaResult,
    }
  }

  completion(
    options: LlamaCompletionOptions,
    callback?: (token: LlamaCompletionToken) => void,
  ): Promise<LlamaCompletionResult> {
    const { messages, media_paths = options.media_paths } =
      this._formatMediaChat(options.messages)
    return this.ctx.completion(
      {
        ...options,
        messages,
        media_paths: options.media_paths || media_paths,
      },
      callback || (() => {}),
    )
  }

  stopCompletion(): void {
    return this.ctx.stopCompletion()
  }

  tokenize(
    text: string,
    { media_paths }: { media_paths?: string[] } = {},
  ): Promise<TokenizeResult> {
    return this.ctx.tokenize(text, media_paths)
  }

  detokenize(tokens: number[]): Promise<string> {
    return this.ctx.detokenize(tokens)
  }

  embedding(text: string, params?: { embd_normalize?: number }): Promise<EmbeddingResult> {
    return this.ctx.embedding(text, params)
  }

  rerank(
    query: string,
    documents: string[],
    params?: RerankParams,
  ): Promise<Array<RerankResult & { document: string }>> {
    return this.ctx
      .rerank(query, documents, params)
      .then((results: RerankResult[]) => {
        // Sort by score descending and add document text for convenience
        return results
          .map((result: RerankResult) => ({
            ...result,
            document: documents[result.index],
          }))
          .sort(
            (
              a: RerankResult & { document: string },
              b: RerankResult & { document: string },
            ) => b.score - a.score,
          )
      })
  }

  saveSession(path: string): Promise<void> {
    return this.ctx.saveSession(path)
  }

  loadSession(path: string): Promise<void> {
    return this.ctx.loadSession(path)
  }

  release(): Promise<void> {
    return this.ctx.release()
  }

  applyLoraAdapters(adapters: { path: string; scaled: number }[]): void {
    return this.ctx.applyLoraAdapters(adapters)
  }

  removeLoraAdapters(): void {
    this.ctx.removeLoraAdapters()
  }

  getLoadedLoraAdapters(): { path: string; scaled: number }[] {
    return this.ctx.getLoadedLoraAdapters()
  }

  initMultimodal(options: { path: string; use_gpu?: boolean }): boolean {
    return this.ctx.initMultimodal(options)
  }

  isMultimodalEnabled(): boolean {
    return this.ctx.isMultimodalEnabled()
  }

  releaseMultimodal(): void {
    this.ctx.releaseMultimodal()
  }

  getMultimodalSupport(): {
    vision: boolean
    audio: boolean
  } {
    return this.ctx.getMultimodalSupport()
  }

  initVocoder(options: { path: string; n_batch?: number }): boolean {
    return this.ctx.initVocoder(options)
  }

  releaseVocoder(): void {
    this.ctx.releaseVocoder()
  }

  isVocoderEnabled(): boolean {
    return this.ctx.isVocoderEnabled()
  }

  getFormattedAudioCompletion(
    speaker: string | null,
    text: string,
  ): {
    prompt: string
    grammar?: string
  } {
    return this.ctx.getFormattedAudioCompletion(speaker, text)
  }

  getAudioCompletionGuideTokens(text: string): Int32Array {
    return this.ctx.getAudioCompletionGuideTokens(text)
  }

  decodeAudioTokens(tokens: number[] | Int32Array): Promise<Float32Array> {
    return this.ctx.decodeAudioTokens(tokens)
  }
}

export const loadModel = async (
  options: LlamaModelOptionsExtended,
): Promise<LlamaContextWrapper> => {
  const variant = options.lib_variant ?? 'default'
  mods[variant] ??= await loadModule(options.lib_variant)
  refreshNativeLogSetup()

  const nativeCtx = new mods[variant].LlamaContext(options)
  return new LlamaContextWrapper(nativeCtx)
}

export const initLlama = loadModule

const modelInfoSkip = [
  // Large fields
  'tokenizer.ggml.tokens',
  'tokenizer.ggml.token_type',
  'tokenizer.ggml.merges',
  'tokenizer.ggml.scores',
]

export const loadLlamaModelInfo = async (
  path: string,
): Promise<GGUFModelInfo> => {
  const variant = 'default'
  mods[variant] ??= await loadModule(variant)
  refreshNativeLogSetup()
  return mods[variant].LlamaContext.loadModelInfo(path, modelInfoSkip)
}

export const BuildInfo = {
  number: BUILD_NUMBER,
  commit: BUILD_COMMIT,
}
