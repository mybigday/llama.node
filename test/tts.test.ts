import path from 'path'
import waitForExpect from 'wait-for-expect'
import { loadModel, loadLlamaModelInfo, toggleNativeLog, addNativeLogListener } from '../lib'

const OUTETTS_GRAMMAR = `\
root       ::= NL? content* audioEnd NL eos?

content ::= wordAudioBlock | emotionBlock

wordAudioBlock ::= punctuation? WORD punctuation? codeBlock space NL

codeBlock ::= TIME CODE{1,144}

emotionBlock ::= emotionStart TEXT emotionEnd space NL

TEXT ::= [A-Za-z0-9 .,?!]+

bos    ::= "<|im_start|>"
eos      ::= "<|im_end|>"
emotionStart ::= "<|emotion_start|>"
emotionEnd ::= "<|emotion_end|>"
audioEnd   ::= "<|audio_end|>"
space      ::= "<|space|>"

WORD       ::= [A-Za-z]+
NL         ::= "\\n"

TIME  ::= "<|t_" DECIMAL "|>"

CODE    ::= "<|" DIGITS "|>"

DIGITS     ::= [0-9]+
DECIMAL    ::= [0-9]+ "." [0-9]+

punctuation ::= "<|" [a-z_]+ "|>"
`

test.skip('TTS', async () => {
  const model = await loadModel({
    model: path.resolve(__dirname, './OuteTTS-0.3-500M-Q4_K_M.gguf'),
    n_ctx: 8192,
    n_batch: 8192,
    n_ubatch: 128,
    n_threads: 4,
    n_gpu_layers: 99,
    ctx_shift: false,
  })
  await model.initVocoder({ path: path.resolve(__dirname, './WavTokenizer.gguf'), n_batch: 2048 })
  expect(model.isVocoderEnabled()).toBe(true)
  const prompt = model.getFormattedAudioCompletion(null, 'Hello, world')
  expect(prompt).toMatchSnapshot('formatted')
  const tokens = model.getAudioCompletionGuideTokens('Hello, world!')
  expect(tokens).toMatchSnapshot('guide tokens')
  const result = await model.completion({
    prompt,
    temperature: 0.2,
    penalty_repeat: 1.2,
    penalty_last_n: 64,
    guide_tokens: tokens,
    top_k: 40,
    top_p: 0.9,
    min_p: 0.05,
    mirostat_tau: 5,
    mirostat_eta: 0.1,
    mirostat: 0,
    grammar: OUTETTS_GRAMMAR,
    n_predict: 3840,
    stop: ['<|audio_end|>', '<|im_end|>'],
  })
  const { audio_tokens: audioTokens } = result
  console.log(result)
  const audio = await model.decodeAudioTokens(audioTokens!)
  expect(audio.length).toBeGreaterThan(0)
  await model.releaseVocoder()
  expect(model.isVocoderEnabled()).toBe(false)
  await model.release()
}, 1e8)
