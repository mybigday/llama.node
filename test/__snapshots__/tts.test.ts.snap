// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TTS: formatted 1`] = `
"<|im_start|>
<|text_start|>the<|space|>overall<|space|>package<|space|>from<|space|>just<|space|>two<|space|>people<|space|>is<|space|>pretty<|space|>remarkable<|space|>sure<|space|>i<|space|>have<|space|>some<|space|>critiques<|space|>about<|space|>some<|space|>of<|space|>the<|space|>gameplay<|space|>aspects<|space|>but<|space|>its<|space|>still<|space|>really<|space|>enjoyable<|space|>and<|space|>it<|space|>looks<|space|>lovely<|space|>hello<|space|>world<|text_end|>
<|audio_start|>
the<|t_0.08|><|257|><|740|><|636|><|913|><|788|><|1703|><|space|>
overall<|t_0.36|><|127|><|201|><|191|><|774|><|700|><|532|><|1056|><|557|><|798|><|298|><|1741|><|747|><|1662|><|1617|><|1702|><|1527|><|368|><|1588|><|1049|><|1008|><|1625|><|747|><|1576|><|728|><|1019|><|1696|><|1765|><|space|>
package<|t_0.56|><|935|><|584|><|1319|><|627|><|1016|><|1491|><|1344|><|1117|><|1526|><|1040|><|239|><|1435|><|951|><|498|><|723|><|1180|><|535|><|789|><|1649|><|1637|><|78|><|465|><|1668|><|901|><|595|><|1675|><|117|><|1009|><|1667|><|320|><|840|><|79|><|507|><|1762|><|1508|><|1228|><|1768|><|802|><|1450|><|1457|><|232|><|639|><|space|>
from<|t_0.19|><|604|><|782|><|1682|><|872|><|1532|><|1600|><|1036|><|1761|><|647|><|1554|><|1371|><|653|><|1595|><|950|><|space|>
just<|t_0.25|><|1782|><|1670|><|317|><|786|><|1748|><|631|><|599|><|1155|><|1364|><|1524|><|36|><|1591|><|889|><|1535|><|541|><|440|><|1532|><|50|><|870|><|space|>
two<|t_0.24|><|1681|><|1510|><|673|><|799|><|805|><|1342|><|330|><|519|><|62|><|640|><|1138|><|565|><|1552|><|1497|><|1552|><|572|><|1715|><|1732|><|space|>
people<|t_0.39|><|593|><|274|><|136|><|740|><|691|><|633|><|1484|><|1061|><|1138|><|1485|><|344|><|428|><|397|><|1562|><|645|><|917|><|1035|><|1449|><|1669|><|487|><|442|><|1484|><|1329|><|1832|><|1704|><|600|><|761|><|653|><|269|><|space|>
is<|t_0.16|><|566|><|583|><|1755|><|646|><|1337|><|709|><|802|><|1008|><|485|><|1583|><|652|><|10|><|space|>
pretty<|t_0.32|><|1818|><|1747|><|692|><|733|><|1010|><|534|><|406|><|1697|><|1053|><|1521|><|1355|><|1274|><|816|><|1398|><|211|><|1218|><|817|><|1472|><|1703|><|686|><|13|><|822|><|445|><|1068|><|space|>
remarkable<|t_0.68|><|230|><|1048|><|1705|><|355|><|706|><|1149|><|1535|><|1787|><|1356|><|1396|><|835|><|1583|><|486|><|1249|><|286|><|937|><|1076|><|1150|><|614|><|42|><|1058|><|705|><|681|><|798|><|934|><|490|><|514|><|1399|><|572|><|1446|><|1703|><|1346|><|1040|><|1426|><|1304|><|664|><|171|><|1530|><|625|><|64|><|1708|><|1830|><|1030|><|443|><|1509|><|1063|><|1605|><|1785|><|721|><|1440|><|923|><|space|>
sure<|t_0.36|><|792|><|1780|><|923|><|1640|><|265|><|261|><|1525|><|567|><|1491|><|1250|><|1730|><|362|><|919|><|1766|><|543|><|1|><|333|><|113|><|970|><|252|><|1606|><|133|><|302|><|1810|><|1046|><|1190|><|1675|><|space|>
i<|t_0.08|><|123|><|439|><|1074|><|705|><|1799|><|637|><|space|>
have<|t_0.16|><|1509|><|599|><|518|><|1170|><|552|><|1029|><|1267|><|864|><|419|><|143|><|1061|><|0|><|space|>
some<|t_0.16|><|619|><|400|><|1270|><|62|><|1370|><|1832|><|917|><|1661|><|167|><|269|><|1366|><|1508|><|space|>
critiques<|t_0.60|><|559|><|584|><|1163|><|1129|><|1313|><|1728|><|721|><|1146|><|1093|><|577|><|928|><|27|><|630|><|1080|><|1346|><|1337|><|320|><|1382|><|1175|><|1682|><|1556|><|990|><|1683|><|860|><|1721|><|110|><|786|><|376|><|1085|><|756|><|1523|><|234|><|1334|><|1506|><|1578|><|659|><|612|><|1108|><|1466|><|1647|><|308|><|1470|><|746|><|556|><|1061|><|space|>
about<|t_0.29|><|26|><|1649|><|545|><|1367|><|1263|><|1728|><|450|><|859|><|1434|><|497|><|1220|><|1285|><|179|><|755|><|1154|><|779|><|179|><|1229|><|1213|><|922|><|1774|><|1408|><|space|>
some<|t_0.23|><|986|><|28|><|1649|><|778|><|858|><|1519|><|1|><|18|><|26|><|1042|><|1174|><|1309|><|1499|><|1712|><|1692|><|1516|><|1574|><|space|>
of<|t_0.07|><|197|><|716|><|1039|><|1662|><|64|><|space|>
the<|t_0.08|><|1811|><|1568|><|569|><|886|><|1025|><|1374|><|space|>
gameplay<|t_0.48|><|1269|><|1092|><|933|><|1362|><|1762|><|1700|><|1675|><|215|><|781|><|1086|><|461|><|838|><|1022|><|759|><|649|><|1416|><|1004|><|551|><|909|><|787|><|343|><|830|><|1391|><|1040|><|1622|><|1779|><|1360|><|1231|><|1187|><|1317|><|76|><|997|><|989|><|978|><|737|><|189|><|space|>
aspects<|t_0.56|><|1423|><|797|><|1316|><|1222|><|147|><|719|><|1347|><|386|><|1390|><|1558|><|154|><|440|><|634|><|592|><|1097|><|1718|><|712|><|763|><|1118|><|1721|><|1311|><|868|><|580|><|362|><|1435|><|868|><|247|><|221|><|886|><|1145|><|1274|><|1284|><|457|><|1043|><|1459|><|1818|><|62|><|599|><|1035|><|62|><|1649|><|778|><|space|>
but<|t_0.20|><|780|><|1825|><|1681|><|1007|><|861|><|710|><|702|><|939|><|1669|><|1491|><|613|><|1739|><|823|><|1469|><|648|><|space|>
its<|t_0.09|><|92|><|688|><|1623|><|962|><|1670|><|527|><|599|><|space|>
still<|t_0.27|><|636|><|10|><|1217|><|344|><|713|><|957|><|823|><|154|><|1649|><|1286|><|508|><|214|><|1760|><|1250|><|456|><|1352|><|1368|><|921|><|615|><|5|><|space|>
really<|t_0.36|><|55|><|420|><|1008|><|1659|><|27|><|644|><|1266|><|617|><|761|><|1712|><|109|><|1465|><|1587|><|503|><|1541|><|619|><|197|><|1019|><|817|><|269|><|377|><|362|><|1381|><|507|><|1488|><|4|><|1695|><|space|>
enjoyable<|t_0.49|><|678|><|501|><|864|><|319|><|288|><|1472|><|1341|><|686|><|562|><|1463|><|619|><|1563|><|471|><|911|><|730|><|1811|><|1006|><|520|><|861|><|1274|><|125|><|1431|><|638|><|621|><|153|><|876|><|1770|><|437|><|987|><|1653|><|1109|><|898|><|1285|><|80|><|593|><|1709|><|843|><|space|>
and<|t_0.15|><|1285|><|987|><|303|><|1037|><|730|><|1164|><|502|><|120|><|1737|><|1655|><|1318|><|space|>
it<|t_0.09|><|848|><|1366|><|395|><|1601|><|1513|><|593|><|1302|><|space|>
looks<|t_0.27|><|1281|><|1266|><|1755|><|572|><|248|><|1751|><|1257|><|695|><|1380|><|457|><|659|><|585|><|1315|><|1105|><|1776|><|736|><|24|><|736|><|654|><|1027|><|space|>
lovely<|t_0.56|><|634|><|596|><|1766|><|1556|><|1306|><|1285|><|1481|><|1721|><|1123|><|438|><|1246|><|1251|><|795|><|659|><|1381|><|1658|><|217|><|1772|><|562|><|952|><|107|><|1129|><|1112|><|467|><|550|><|1079|><|840|><|1615|><|1469|><|1380|><|168|><|917|><|836|><|1827|><|437|><|583|><|67|><|595|><|1087|><|1646|><|1493|><|1677|><|space|>
"
`;

exports[`TTS: guide tokens 1`] = `
Int32Array [
  198,
  14990,
  14615,
  151668,
]
`;
